Support patient write scope

imranq2 opened the pull request:
This PR adds a check for the patient scope writes.
There are unit tests added to check the behavior
I pass the requestInfo into the database writers now.  Lots of files changed for this.
Added a preSave handler that checks for patient scope writes and throws a Forbidden error if so.
Also updated from node 20.10 to 20.11.1 to stay current.
Upgraded the github actions to node 20 to stop warnings from github.

cdukebwell reviewed (commented):
Does the process to check if a user has the ability to write a resource look the the:
- existing resource to see if the records' patient reference is one of the user patient id(s) 
- new resource to see if the records' patient reference is one of the user patient id(s)

imranq2 commented:
> Does the process to check if a user has the ability to write a resource look the the:
> 
> * existing resource to see if the records' patient reference is one of the user patient id(s)
> * new resource to see if the records' patient reference is one of the user patient id(s)

Yes, added a unit test to confirm behavior.

imranq2 commented:
> If a bundle is sent where some resources get updated and others dont due to patient id restrictions should the entire request fail? if it partially fails should it send a 200 or a 403?

Fixed it and added a unit test verify.
